src/test/unit/pid_unittest.cc

```diff
@@ -0,0 +1,464 @@
+    DISABLE_ARMING_FLAG(ARMED);
+
+    setDefaultTestSettings();
+    for (int axis = FD_ROLL; axis <= FD_YAW; axis++) {
+        pidData[axis].P = 0;
+        pidData[axis].I = 0;
+        pidData[axis].D = 0;
+        pidData[axis].Sum = 0;
+        simulatedSetpointRate[axis] = 0;
+        simulatedRcDeflection[axis] = 0;
+        gyro.gyroADCf[axis] = 0;
+    }
+    attitude.values.roll = 0;
+    attitude.values.pitch = 0;
+    attitude.values.yaw = 0;
+
+    flightModeFlags = 0;
+    pidInit(pidProfile);
+
+    // Run pidloop for a while after reset
+    for (int loop = 0; loop < 20; loop++) {
+        pidController(pidProfile, &rollAndPitchTrims, currentTestTime());
+    }
+}
+
+void setStickPosition(int axis, float stickRatio) {
+    simulatedSetpointRate[axis] = 1998.0f * stickRatio;
+    simulatedRcDeflection[axis] = stickRatio;
+}
+
+TEST(pidControllerTest, testInitialisation)
+{
+    resetTest();
+
+    // In initial state PIDsums should be 0
+    for (int axis = 0; axis <= FD_YAW; axis++) {
+        EXPECT_FLOAT_EQ(0, pidData[axis].P);
+        EXPECT_FLOAT_EQ(0, pidData[axis].I);
+        EXPECT_FLOAT_EQ(0, pidData[axis].D);
+    }
+}
+
+TEST(pidControllerTest, testStabilisationDisabled) {
+    ENABLE_ARMING_FLAG(ARMED);
+    // Run few loops to make sure there is no error building up when stabilisation disabled
+
+    for (int loop = 0; loop < 10; loop++) {
+        pidController(pidProfile, &rollAndPitchTrims, currentTestTime());
+
+        // PID controller should not do anything, while stabilisation disabled
+        EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].P);
+        EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].P);
+        EXPECT_FLOAT_EQ(0, pidData[FD_YAW].P);
+        EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].I);
+        EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].I);
+        EXPECT_FLOAT_EQ(0, pidData[FD_YAW].I);
+        EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].D);
+        EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].D);
+        EXPECT_FLOAT_EQ(0, pidData[FD_YAW].D);
+    }
+}
+
+TEST(pidControllerTest, testPidLoop) {
+    // Make sure to start with fresh values
+    resetTest();
+    ENABLE_ARMING_FLAG(ARMED);
+    pidStabilisationState(PID_STABILISATION_ON);
+
+    pidController(pidProfile, &rollAndPitchTrims, currentTestTime());
+
+    // Loop 1 - Expecting zero since there is no error
+    EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].P);
+    EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].P);
+    EXPECT_FLOAT_EQ(0, pidData[FD_YAW].P);
+    EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].I);
+    EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].I);
+    EXPECT_FLOAT_EQ(0, pidData[FD_YAW].I);
+    EXPECT_FLOAT_EQ(0, pidData[FD_ROLL].D);
+    EXPECT_FLOAT_EQ(0, pidData[FD_PITCH].D);
+    EXPECT_FLOAT_EQ(0, pidData[FD_YAW].D);
+
+    // Add some rotation on ROLL to generate error
+    gyro.gyroADCf[FD_ROLL] = 100;
+    pidController(pidProfile, &rollAndPitchTrims, currentTestTime());
+
+    // Loop 2 - Expect PID loop reaction to ROLL error
+    EXPECT_FLOAT_EQ(-128.116, pidData[FD_ROLL].P);
```

Comment:
This seems far to fragile. Either quite wide interval should be used (+-10% ?) or simply check that result is in correct direction ...

Reply:
So you suggest to use less strict checking with tolerances? Maybe good idea indeed

Reply:
`EXPECT_NEAR(a, b, abs_tolerance)`

Reply:
These tests have 0 maintenability as long as they use these magic float values.
Not only no one besides you understands their origin and meaning, they are destined to break

Reply:
I think having tolerance of 10% would be good one.

Reply:
```ASSERT_NEAR (expected, actual, absoluteRange);```